src/vs/workbench/contrib/terminalContrib/suggest/browser/lspCompletionProviderAddon.ts

```diff
@@ -0,0 +1,136 @@
+	private _textVirtualModel: IReference<IResolvedTextEditorModel>;
+	private _lspTerminalModelContentProvider: LspTerminalModelContentProvider;
+
+	constructor(
+		provider: CompletionItemProvider,
+		textVirtualModel: IReference<IResolvedTextEditorModel>,
+		lspTerminalModelContentProvider: LspTerminalModelContentProvider,
+	) {
+		super();
+		this._provider = provider;
+		this._textVirtualModel = textVirtualModel;
+		this._lspTerminalModelContentProvider = lspTerminalModelContentProvider;
+		this.triggerCharacters = provider.triggerCharacters ? [...provider.triggerCharacters, ' '] : [' '];
+	}
+
+	activate(terminal: Terminal): void {
+		// console.log('activate');
+	}
+
+	async provideCompletions(value: string, cursorPosition: number, allowFallbackCompletions: false, token: CancellationToken): Promise<ITerminalCompletion[] | TerminalCompletionList<ITerminalCompletion> | undefined> {
+
+		// Apply edit for non-executed current commandline --> Pretend we are typing in the real-document.
+		this._lspTerminalModelContentProvider.trackPromptInputToVirtualFile(value);
+
+		const textBeforeCursor = value.substring(0, cursorPosition);
+		const lines = textBeforeCursor.split('\n');
+		const column = lines[lines.length - 1].length + 1;
+
+		// Get line from virtualDocument, not from terminal
+		const lineNum = this._textVirtualModel.object.textEditorModel.getLineCount();
+		const positionVirtualDocument = new Position(lineNum, column);
+
+
+		// TODO: Scan back to start of nearest word like other providers? Is this needed for `ILanguageFeaturesService`?
+		const completions: ITerminalCompletion[] = [];
+		if (this._provider && this._provider._debugDisplayName !== 'wordbasedCompletions') {
+
+			const result = await this._provider.provideCompletionItems(this._textVirtualModel.object.textEditorModel, positionVirtualDocument, { triggerKind: CompletionTriggerKind.TriggerCharacter }, token);
+
+			completions.push(...(result?.suggestions || []).map((e: any) => {
+				// TODO: Support more terminalCompletionItemKind for different LSP providers.
+				const convertedKind = e.kind ? mapLspKindToTerminalKind(e.kind) : TerminalCompletionItemKind.Method;
+				const completionItemTemp = createCompletionItemPython(cursorPosition, textBeforeCursor, convertedKind, 'lspCompletionItem', undefined);
+
+				return {
+					label: e.insertText,
+					provider: `lsp:${this._provider._debugDisplayName}`,
+					detail: e.detail,
+					kind: convertedKind,
+					replacementIndex: completionItemTemp.replacementIndex,
+					replacementLength: completionItemTemp.replacementLength,
+				};
+			}));
+		}
+
+		return completions;
+	}
+}
+
+export function createCompletionItemPython(cursorPosition: number, prefix: string, kind: TerminalCompletionItemKind, label: string | CompletionItemLabel, detail: string | undefined): TerminalCompletionItem {
+	const endsWithDot = prefix.endsWith('.');
+	const endsWithSpace = prefix.endsWith(' ');
+
+	if (endsWithSpace) {
```

Comment:
some comments with examples of what each of these cases might be would be good here so it's easy to understand years from now 😄

Reply:
For sure! Tried here: [6a897d9](https://github.com/microsoft/vscode/pull/243528/commits/6a897d9136f20badcc9bf96a52dbddeb5106b2dc)